nibabel/streamlines/tractogram.py

```diff
@@ -0,0 +1,558 @@
+        """ Returns the affine bringing this tractogram to RAS+mm. """
+        return self._affine_to_rasmm.copy()
+
+    def __iter__(self):
+        for i in range(len(self.streamlines)):
+            yield self[i]
+
+    def __getitem__(self, idx):
+        pts = self.streamlines[idx]
+
+        data_per_streamline = {}
+        for key in self.data_per_streamline:
+            data_per_streamline[key] = self.data_per_streamline[key][idx]
+
+        data_per_point = {}
+        for key in self.data_per_point:
+            data_per_point[key] = self.data_per_point[key][idx]
+
+        if isinstance(idx, (int, np.integer)):
+            return TractogramItem(pts, data_per_streamline, data_per_point)
+
+        return Tractogram(pts, data_per_streamline, data_per_point)
+
+    def __len__(self):
+        return len(self.streamlines)
+
+    def copy(self):
+        """ Returns a copy of this :class:`Tractogram` object. """
+        return copy.deepcopy(self)
+
+    def apply_affine(self, affine, lazy=False):
+        """ Applies an affine transformation on the points of each streamline.
+
+        If `lazy` is not specified, this is performed *in-place*.
+
+        Parameters
+        ----------
+        affine : ndarray shape (4, 4)
+            Transformation that will be applied to every streamline.
+        lazy_load : {False, True}, optional
+            If True, streamlines are *not* transformed in-place and a
+            :class:`LazyTractogram` object is returned. Otherwise, streamlines
+            are modified in-place.
+
+        Returns
+        -------
+        tractogram : :class:`Tractogram` or :class:`LazyTractogram` object
+            Tractogram where the streamlines have been transformed according
+            to the given affine transformation. If the `lazy` option is true,
+            it returns a :class:`LazyTractogram` object, otherwise it returns a
+            reference to this :class:`Tractogram` object with updated
+            streamlines.
+        """
+        if lazy:
+            lazy_tractogram = LazyTractogram.from_tractogram(self)
+            lazy_tractogram.apply_affine(affine)
+            return lazy_tractogram
+
+        if len(self.streamlines) == 0:
+            return self
+
+        BUFFER_SIZE = 10000000  # About 128 Mb since pts shape is 3.
+        for start in range(0, len(self.streamlines._data), BUFFER_SIZE):
+            end = start + BUFFER_SIZE
+            pts = self.streamlines._data[start:end]
+            self.streamlines._data[start:end] = apply_affine(affine, pts)
+
+        # Update the affine that brings back the streamlines to RASmm.
+        self._affine_to_rasmm = np.dot(self._affine_to_rasmm,
+                                       np.linalg.inv(affine))
+
+        return self
+
+
+class LazyTractogram(Tractogram):
+    """ Class containing information about streamlines.
+
+    Tractogram objects have four main properties: `header`, `streamlines`,
```

Comment:
An explanation here of the difference between `Tractogram` and `LazyTractogram` ?